scalalib/src/mill/scalalib/JavaModule.scala

```diff
@@ -847,10 +847,52 @@ trait JavaModule
    */
   def generatedSources: T[Seq[PathRef]] = Task { Seq.empty[PathRef] }
 
+  /**
+   * A set of tasks producing generated code. The difference between this and
+   * [[generatedSources]] lies in the fact that IDE-related tasks
+   * (BSP/Bloop/GenIdea) will not run the tasks, but will use them to statically
+   * determine the location of the source code they will produce by looking up their
+   * [[Task#generatedSourceRoots]], knowing that these roots will be present under
+   * the dest directories once the task is run.
+   */
+  def deferredSourceGenerators: Seq[NamedTask[Unit]] = Seq.empty
+
+  private def deferredSourceRoots(task: NamedTask[Unit])(workspace: os.Path): Seq[os.Path] = {
+    val dest = mill
+      .eval
+      .EvaluatorPaths
+      .resolveDestPaths(workspace / "out", task)
+      .dest
+    val explicitSourceRoots = task.generatedSourceRoots
+    if (explicitSourceRoots.isEmpty) Seq(dest)
+    else explicitSourceRoots.map(subPath => dest / subPath)
+  }
+
+  /**
+   * Computes the list of source roots that will be produced by [[deferredSourceGenerators]] without
+   * actually running the generators in question.
+   */
+  def allDeferredSourceRoots: T[Seq[PathRef]] = T {
+    val ws = T.workspace
+    deferredSourceGenerators.flatMap(t => deferredSourceRoots(t)(ws)).map(PathRef(_))
+  }
+
+  /**
+   * Runs the lazy source generators, returning references to the expanded generated source roots
+   * they are supposed to have written code in.
+   */
+  final def deferredGeneratedSources: T[Seq[PathRef]] = T {
+    val ws = T.workspace
+    T.sequence {
+      deferredSourceGenerators.map { t => t.map((_: Unit) => deferredSourceRoots(t)) }
+    }().flatMap(_.apply(ws)).map(PathRef(_))
+  }
+
   /**
    * The folders containing all source files fed into the compiler
    */
-  def allSources: T[Seq[PathRef]] = Task { sources() ++ generatedSources() }
+  def allSources: T[Seq[PathRef]] =
+    Task { sources() ++ generatedSources() ++ deferredGeneratedSources() }
```

Comment:
Keeps the current (0.12.x) behaviour of `generatedSources` to avoid breaking the ecosystem